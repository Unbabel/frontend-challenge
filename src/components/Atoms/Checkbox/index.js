import React, { useState } from 'react'
import styled from 'styled-components'
import { prop } from 'styled-tools'


const CheckboxContainer = styled.div`
  display: inline-block;
  vertical-align: middle;
  margin: .25rem;
`

const Icon = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`
// Hide checkbox visually but remain accessible to screen readers.
// Source: https://polished.js.org/docs/#hidevisually
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  
`

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 16px;
  height: 16px;
  background: ${props => (props.checked ? '#ffffff' : '#859EFF')}
  border-radius: 3px;
  border: 3px solid #859EFF;
  transition: all 150ms;
  cursor:pointer;
  line-height:1rem;

  ${HiddenCheckbox}:focus + & {
    box-shadow: 0 0 0 3px pink;
  }

  ${Icon} {
    visibility: ${props => (props.checked ? 'hidden' : 'visible')}
  }
`

const Checkbox = ({ ...props }) => {
  const [checkedItem, setCheckedItem] = useState({ value: false })

  const handleChange = (event) => {
    setCheckedItem(!checkedItem)
    console.log('checkedItem: ', checkedItem)
  }
  return (
    <CheckboxContainer>
      <HiddenCheckbox {...props} />
      <StyledCheckbox {...props} checked={checkedItem} onClick={handleChange}>
        <Icon viewBox="0 0 24 24">
          <polyline points="20 6 9 17 4 12" />
        </Icon>
      </StyledCheckbox>
    </CheckboxContainer>
  )
}
export default Checkbox
