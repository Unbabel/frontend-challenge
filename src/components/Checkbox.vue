<template>
    <label class="custom-checkbox">
        <input name="custom checkbox" type="checkbox">
        <span class="checkmark"></span>
    </label>
</template>

<script>
export default {
    name: 'Checkbox'
}
</script>

<style lang="scss" scoped>
    .custom-checkbox {
        position: relative;
        border: 1px solid #859EFF;
        width: 16px;
        height: 16px;
        border-radius: 1px;
        &:hover {
            cursor: pointer;
        }
        input {
            opacity: 0;
            &:checked ~ .checkmark {
                background-color: #859EFF;
            }
            &:checked ~ .checkmark:after {
                opacity: 1;
            }
        }
        .checkmark {
            display: block;

            position: absolute;
            top: 0;
            left: 0;
            height: 15px;
            width: 15px;
            background-color: transparent;
            &:after {
                content: "";
                border: 2px solid #fff;
                border-left: 0;
                border-top: 0;
                height: 9px;
                width: 4px;
                left: 4px;
                top: 0;
                border-radius: 1px;
                opacity: 0;
                position: absolute;
                transform: rotate(45deg);
                transition: opacity 0.2s ease-in-out;
            }
        }
    }
</style>


