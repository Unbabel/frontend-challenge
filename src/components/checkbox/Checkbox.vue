<script setup>
	import { computed } from "vue";
	const { id } = defineProps({
		id: {
			type: String,
			required: true,
		},
	});

	const checkboxId = computed(() => "checkbox-" + id);
</script>

<template>
	<div class="checkbox-div">
		<input
			name="cssCheckbox"
			:id="checkboxId"
			type="checkbox"
			class="css-checkbox"
			aria-checked
			checked
		/>
		<label :for="checkboxId"></label>
	</div>
</template>

<style scoped>
	.checkbox-div {
		height: 1rem;
		width: 1rem;
	}

	.css-checkbox {
		position: absolute;
		overflow: hidden;
		clip: rect(0 0 0 0);
		height: 0.0625rem;
		width: 0.0625rem;
		padding: 0;
		border: 0;
	}

	.css-checkbox + label {
		position: relative;
		font-size: 1rem;
		cursor: pointer;
		display: inline-flex;
		align-items: center;
	}

	.css-checkbox + label::before {
		content: " ";
		vertical-align: middle;
		height: 0.75rem;
		width: 0.75rem;
		border-width: 0.125rem;
		border-style: solid;
		border-color: var(--palette-purple);
		border-radius: 0.125rem;
		box-shadow: none;
	}

	.css-checkbox:checked + label::after {
		content: " ";
		background-image: url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA1MTIgNTEyIj48cGF0aCBmaWxsPSIjZmZmZmZmIiBkPSJNMTczLjg5OCA0MzkuNDA0bC0xNjYuNC0xNjYuNGMtOS45OTctOS45OTctOS45OTctMjYuMjA2IDAtMzYuMjA0bDM2LjIwMy0zNi4yMDRjOS45OTctOS45OTggMjYuMjA3LTkuOTk4IDM2LjIwNCAwTDE5MiAzMTIuNjkgNDMyLjA5NSA3Mi41OTZjOS45OTctOS45OTcgMjYuMjA3LTkuOTk3IDM2LjIwNCAwbDM2LjIwMyAzNi4yMDRjOS45OTcgOS45OTcgOS45OTcgMjYuMjA2IDAgMzYuMjA0bC0yOTQuNCAyOTQuNDAxYy05Ljk5OCA5Ljk5Ny0yNi4yMDcgOS45OTctMzYuMjA0LS4wMDF6Ii8+PC9zdmc+");
		background-repeat: no-repeat;
		background-size: 0.6rem 1rem;
		background-position: center center;
		position: absolute;
		display: flex;
		justify-content: center;
		align-items: center;
		text-align: center;
		background-color: var(--palette-purple);
		font-size: 1rem;
		height: 0.875rem;
		width: 0.985rem;
		border-radius: 0.125rem;
	}
</style>
