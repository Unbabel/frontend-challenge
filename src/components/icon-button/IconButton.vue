<script setup>
	const emit = defineEmits(["onClickCallback"]);

	defineProps({
		icon: {
			type: String,
			required: true,
		},
	});

	const onClickHandler = () => emit("onClickCallback");
</script>

<template>
	<button
		class="icon-button"
		@click="onClickHandler"
		data-test="icon-button"
		aria-label="Icon Button"
	>
		<img :src="icon" alt="button-icon" />
	</button>
</template>

<style scoped>
	.icon-button {
		background-color: transparent;
		border: none;
	}

	.icon-button:hover {
		cursor: pointer;
	}
</style>
