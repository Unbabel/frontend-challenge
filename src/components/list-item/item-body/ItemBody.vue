<script setup>
	// Components
	import Input from "../../input/Input.vue";

	defineProps({
		content: {
			type: String,
		},
	});

	const emit = defineEmits(["onChangeCallback"]);

	const onChangeHandler = (value) => emit("onChangeCallback", value);
</script>

<template>
	<Input
		customStyles="content-styles"
		:value="content"
		:onChangeCallback="onChangeHandler"
		placeholder="Insert content..."
	/>
</template>

<style scoped>
	.content-styles {
		font-family: "Open Sans";
		font-weight: 400;
		color: var(--palette-grey);
		text-justify: distribute;
		font-size: 1.125rem;
	}
</style>
