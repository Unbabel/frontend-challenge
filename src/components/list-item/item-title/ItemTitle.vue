<script setup>
	// Components
	import Input from "../../input/Input.vue";

	defineProps({
		text: {
			type: String,
		},
	});

	const emit = defineEmits(["onChangeCallback"]);

	const onChangeHandler = (value) => emit("onChangeCallback", value);
</script>

<template>
	<Input
		customStyles="title-styles"
		:value="text"
		:onChangeCallback="onChangeHandler"
		placeholder="Insert title..."
	/>
</template>

<style scoped>
	.title-styles {
		font-family: "Montserrat";
		font-weight: 600;
		color: var(--palette-light-grey);
		max-height: 1.25rem;
		font-size: 1rem;
	}
</style>
