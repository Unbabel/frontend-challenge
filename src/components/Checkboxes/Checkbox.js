import * as React from 'react'
import '../../resources/styles/checkbox.scss'

const Checkbox = ({ onClick,activeCheckbox }) => {
    return (<React.Fragment>
        <div className="checkbox" >
            <label
                onClick={onClick}
                className={`label-checkbox ${activeCheckbox ? "checked" : ""}`}
            />
        </div>
    </React.Fragment>
    )
}
export default Checkbox;