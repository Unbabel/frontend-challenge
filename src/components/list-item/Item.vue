<script>
	// Store
	import { mapActions } from "vuex";

	// Components
	import ItemTitle from "./item-title/ItemTitle.vue";
	import ItemBody from "./item-body/ItemBody.vue";
	import Checkbox from "../checkbox/Checkbox.vue";
	import IconButton from "../icon-button/IconButton.vue";

	export default {
		components: {
			ItemTitle,
			ItemBody,
			Checkbox,
			IconButton,
		},
		props: {
			itemId: {
				type: String,
				required: true,
			},
			title: {
				type: String,
			},
			content: {
				type: String,
			},
		},
		methods: {
			...mapActions({
				deleteListItem: "deleteListItem",
				editListItem: "editListItem",
			}),
			onDeleteCallback() {
				this.deleteListItem(this.itemId);
			},
			editListItemTitle(value) {
				this.editListItem({
					itemId: this.itemId,
					key: "voice",
					value: value,
				});
			},
			editListItemBody(value) {
				this.editListItem({
					itemId: this.itemId,
					key: "text",
					value: value,
				});
			},
		},
		data() {
			return {
				showDeleteBtn: false,
			};
		},
	};
</script>

<template>
	<div
		class="item-container"
		@mouseover="showDeleteBtn = true"
		@mouseleave="showDeleteBtn = false"
	>
		<div class="grid-container">
			<Checkbox :id="itemId" />
			<img src="src/assets/person.svg" alt="person-icon" />
			<div class="item-title-wrapper text-margin">
				<ItemTitle
					:text="title"
					@onChangeCallback="editListItemTitle"
				/>
			</div>
			<span
				class="delete-span"
				:class="showDeleteBtn && 'visible'"
				data-test="delete-item-span"
			>
				<IconButton
					icon="src/assets/delete.svg"
					@onClickCallback="onDeleteCallback"
				/>
			</span>
			<div class="empty-cell"><!-- Empty cell --></div>
			<div class="empty-cell"><!-- Empty cell --></div>
			<div class="item-body-wrapper text-margin">
				<ItemBody
					:content="content"
					@onChangeCallback="editListItemBody"
				/>
			</div>
		</div>
	</div>
</template>

<style scoped>
	.grid-container {
		display: grid;
		grid-template-columns: min-content min-content auto min-content;
		width: 100%;
		align-content: center;
		align-items: center;
		column-gap: 1rem;
	}

	.grid-container-row-2 {
		display: grid;
		grid-template-columns: repeat(auto-fill, 10% 10% 80%);
		width: 100%;
		align-content: center;
		align-items: center;
	}

	.item-container {
		display: flex;
		flex-direction: column;
		align-items: flex-start;
	}

	.item-title-wrapper {
		width: 100%;
		padding-top: 0.5rem;
	}

	.visible {
		visibility: visible !important;
	}

	.delete-span {
		width: 100%;
		text-align: end;
		visibility: hidden;
	}

	.text-margin {
		margin-left: -0.5rem;
	}

	@media (max-width: 768px) {
		.empty-cell {
			display: none;
		}

		.item-body-wrapper {
			grid-column-start: 1;
			grid-column-end: 5;
		}
	}

	/* Check if on mobile */
	@media (pointer: none), (pointer: coarse) {
		.delete-span {
			visibility: visible;
		}
	}
</style>
