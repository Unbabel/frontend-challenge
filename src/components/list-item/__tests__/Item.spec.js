import Vuex from "vuex";
import { describe, it, expect, beforeEach, vi } from "vitest";
import { mount, createLocalVue, shallowMount } from "@vue/test-utils";

// Components
import Item from "../Item.vue";
import ItemTitle from "../item-title/ItemTitle.vue";
import ItemBody from "../item-body/ItemBody.vue";
import Checkbox from "../../checkbox/Checkbox.vue";
import IconButton from "../../icon-button/IconButton.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({});

describe("Item", () => {
	let wrapper;
	let propsData = {
		itemId: "1",
		title: "This is some title text",
		content: "This is some content text",
	};

	beforeEach(() => {
		// Arrange
		wrapper = mount(Item, {
			propsData,
			localVue,
			store,
		});
	});

	it("matches snapshot", () => {
		// Assert
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("renders properly", () => {
		// Act
		const itemTitle = wrapper.findComponent(ItemTitle);
		const itemBody = wrapper.findComponent(ItemBody);
		const checkbox = wrapper.findComponent(Checkbox);
		const iconButton = wrapper.findComponent(IconButton);

		// Assert
		expect(itemTitle.exists()).toBe(true);
		expect(itemBody.exists()).toBe(true);
		expect(checkbox.exists()).toBe(true);
		expect(iconButton.exists()).toBe(true);
	});

	it("ItemTitle has correct text", () => {
		// Act
		const itemTitle = wrapper.findComponent(ItemTitle);

		// Assert
		expect(itemTitle.props("text")).toBe("This is some title text");
	});

	it("ItemBody has correct content", () => {
		// Act
		const itemBody = wrapper.findComponent(ItemBody);

		// Assert
		expect(itemBody.props("content")).toBe("This is some content text");
	});

	[
		{ element: "div", clx: "item-container" },
		{ element: "div", clx: "grid-container" },
		{ element: "div", clx: "item-title-wrapper" },
		{ element: "div", clx: "text-margin", cardinality: 2 },
		{ element: "span", clx: "delete-span" },
		{ element: "div", clx: "empty-cell", cardinality: 2 },
		{ element: "div", clx: "item-body-wrapper" },
	].forEach(({ element, clx, cardinality = 1 }) =>
		it("has correct css classes", () => {
			// Act
			const elements = wrapper.findAll(element);
			const filteredElements = elements.filter((el) => el.classes(clx));

			// Assert
			expect(filteredElements).toHaveLength(cardinality);
		})
	);

	it('should have "visible" class on hover', async () => {
		// Act
		const firstDiv = await wrapper.findAll("div").at(0);
		await firstDiv.trigger("mouseover");

		// Assert
		const span = await wrapper.find("span");
		expect(span.classes("visible")).toBe(true);
	});

	it("should call editListItemTitle", async () => {
		// Arrange
		const spy = vi.spyOn(Item.methods, "editListItemTitle");
		wrapper = mount(Item, { propsData, localVue, store });

		// Act
		const itemTitle = wrapper.findComponent(ItemTitle);
		await itemTitle.trigger("input", { value: "Teste" });

		// Assert
		expect(spy).toBeCalled();
	});

	it("should call editListItemBody", async () => {
		// Arrange
		const spy = vi.spyOn(Item.methods, "editListItemBody");
		wrapper = mount(Item, { propsData, localVue, store });

		// Act
		const itemBody = wrapper.findComponent(ItemBody);
		await itemBody.trigger("input", { value: "Teste" });

		// Assert
		expect(spy).toBeCalled();
	});

	it("should call onDeleteCallback", async () => {
		// Arrange
		const spy = vi.spyOn(Item.methods, "onDeleteCallback");
		wrapper = mount(Item, { propsData, localVue, store });

		// Act
		const firstDiv = await wrapper.findAll("div").at(0);
		await firstDiv.trigger("mouseover");

		const iconButton = wrapper.findComponent(IconButton);
		await iconButton.trigger("click");

		// Assert
		expect(spy).toBeCalled();
	});

	it("should call deleteListItem", async () => {
		// Arrange
		const spy = vi.spyOn(Item.methods, "deleteListItem");
		wrapper = mount(Item, { propsData, localVue, store });

		// Act
		const firstDiv = await wrapper.findAll("div").at(0);
		await firstDiv.trigger("mouseover");

		const iconButton = wrapper.findComponent(IconButton);
		await iconButton.trigger("click");

		// Assert
		expect(spy).toBeCalled();
	});
});
