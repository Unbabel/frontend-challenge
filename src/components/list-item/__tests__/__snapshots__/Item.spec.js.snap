// Vitest Snapshot v1

exports[`Item > matches snapshot 1`] = `
"<div data-v-3298ef32=\\"\\" class=\\"item-container\\">
  <div data-v-3298ef32=\\"\\" class=\\"grid-container\\">
    <div data-v-f2e07ba2=\\"\\" data-v-3298ef32=\\"\\" class=\\"checkbox-div\\"><input data-v-f2e07ba2=\\"\\" name=\\"cssCheckbox\\" id=\\"checkbox-1\\" type=\\"checkbox\\" aria-checked=\\"\\" checked=\\"checked\\" class=\\"css-checkbox\\"><label data-v-f2e07ba2=\\"\\" for=\\"checkbox-1\\"></label></div><img data-v-3298ef32=\\"\\" src=\\"src/assets/person.svg\\" alt=\\"person-icon\\">
    <div data-v-3298ef32=\\"\\" class=\\"item-title-wrapper text-margin\\"><textarea data-v-0f426a66=\\"\\" data-v-52cd4611=\\"\\" data-v-3298ef32=\\"\\" placeholder=\\"Insert title...\\" class=\\"title-styles undefined\\" style=\\"height: 0px;\\"></textarea></div><span data-v-3298ef32=\\"\\" data-test=\\"delete-item-span\\" class=\\"delete-span\\"><button data-v-f509f990=\\"\\" data-v-3298ef32=\\"\\" data-test=\\"icon-button\\" aria-label=\\"Icon Button\\" class=\\"icon-button\\"><img data-v-f509f990=\\"\\" src=\\"src/assets/delete.svg\\" alt=\\"button-icon\\"></button></span>
    <div data-v-3298ef32=\\"\\" class=\\"empty-cell\\"></div>
    <div data-v-3298ef32=\\"\\" class=\\"empty-cell\\"></div>
    <div data-v-3298ef32=\\"\\" class=\\"item-body-wrapper text-margin\\"><textarea data-v-0f426a66=\\"\\" data-v-2a797403=\\"\\" data-v-3298ef32=\\"\\" placeholder=\\"Insert content...\\" class=\\"content-styles undefined\\" style=\\"height: 0px;\\"></textarea></div>
  </div>
</div>"
`;
